server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java

```diff
@@ -28,46 +28,52 @@ protected BytesReference newBytesReference(int length) throws IOException {
+    }
+
+    @Override
+    protected BytesReference newBytesReference(byte[] content) throws IOException {
         BytesReference delegate;
         String composite = "composite";
         String paged = "paged";
         String array = "array";
         String type = randomFrom(composite, paged, array);
         if (array.equals(type)) {
-            final BytesStreamOutput out = new BytesStreamOutput(length);
-            for (int i = 0; i < length; i++) {
-                out.writeByte((byte) random().nextInt(1 << 8));
+            final BytesStreamOutput out = new BytesStreamOutput(content.length);
+            for (int i = 0; i < content.length; i++) {
+                out.writeByte(content[i]);
             }
-            assertThat(length, equalTo(out.size()));
-            BytesArray ref = new BytesArray(out.bytes().toBytesRef().bytes, 0, length);
-            assertThat(length, equalTo(ref.length()));
-            assertThat(ref.length(), Matchers.equalTo(length));
+            assertThat(content.length, equalTo(out.size()));
+            BytesArray ref = new BytesArray(out.bytes().toBytesRef().bytes, 0, content.length);
+            assertThat(content.length, equalTo(ref.length()));
+            assertThat(ref.length(), Matchers.equalTo(content.length));
             delegate = ref;
         } else if (paged.equals(type)) {
-            ByteArray byteArray = bigarrays.newByteArray(length);
-            for (int i = 0; i < length; i++) {
-                byteArray.set(i, (byte) random().nextInt(1 << 8));
+            ByteArray byteArray = bigarrays.newByteArray(content.length);
+            for (int i = 0; i < content.length; i++) {
+                byteArray.set(i, content[i]);
             }
-            assertThat(byteArray.size(), Matchers.equalTo((long) length));
-            BytesReference ref = BytesReference.fromByteArray(byteArray, length);
-            assertThat(ref.length(), Matchers.equalTo(length));
+            assertThat(byteArray.size(), Matchers.equalTo((long) content.length));
+            BytesReference ref = BytesReference.fromByteArray(byteArray, content.length);
+            assertThat(ref.length(), Matchers.equalTo(content.length));
             delegate = ref;
         } else {
             assert composite.equals(type);
             List<BytesReference> referenceList = new ArrayList<>();
-            for (int i = 0; i < length;) {
-                int remaining = length - i;
+            for (int i = 0; i < content.length;) {
+                int remaining = content.length - i;
                 int sliceLength = randomIntBetween(1, remaining);
                 ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(sliceLength, bigarrays);
+                int offset = content.length - remaining;
                 for (int j = 0; j < sliceLength; j++) {
-                    out.writeByte((byte) random().nextInt(1 << 8));
+                    out.writeByte(content[offset + j]);
                 }
```

Comment:
This was the only problem in 9d058d1ed18d34afa10f4ff44daefe3f88144d2b I think, it needed to write only the relevant slice of `content`:

```suggestion
                out.writeBytes(content, content.length - remaining, sliceLength);
```